### ISSUE TITLE
Context Help doesn't work for long section names, if Firefox is the default browser

### ISSUE THREAD

**Issue by XLTechie** [nvaccess/nvda#12362]
Note. This is almost surely a Firefox bug (if it can be replicated). However it causes NVDA's Context Help to fail, which is why I'm posting it here.

### Steps to reproduce:

Attempt to open any context help URL from the run dialog, with Firefox set as your default browser.
For example, using the URL `file:///C:/Program Files (x86)/NVDA/documentation/en/userGuide.html`, and appending one of the anchors below.

This is if pasted directly into the dialog, so Windows has to derive the correct executable, similar to how gui.contextHelp does it. If you prepend with "`firefox.exe `", the problem doesn't occur.

### Alternative STR:

1. Press NVDA+h, c, to check for updates.
2. While an update is pending, press F1.

### Actual behavior:

On my system I am finding that:

1. Short anchors, like these, work just fine.
    * `#Introduction`
    * `#GeneralFeatures`
    * `#SystemWideParameters`
    * `#Introduction`
2. Long anchors, such as these, take one to the top of the manual (TOC):
    * `#RunCOMRegistrationFixingTool`
    * `#PortableAndTemporaryCopyRestrictions`
    * `#InstallWithIncompatibleAddons`
    * `#CopyPortableConfigurationToCurrentUserAccount`
    * `#WordAutomaticColumnAndRowHeaderReading`

At first I thought this was because the anchors were over 25 characters long, like `#RunCOMRegistrationFixingTool` (28 characters); but then I found that `#GeneralSettingsCheckForUpdates` (which is 30 characters), works fine.
I now conclude that it relates to long anchors near the end of the document.

### Expected behavior:

All anchors take one to the proper manual section under all circumstances.

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:
Tested a few Alphas.
Currently on alpha-22581,86fe20ff

#### Windows version:
Version 20H2 (OS Build 19042.928)

#### Name and version of other software in use when reproducing the issue:
Firefox 88.0 (64 bit)

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Same, see above.

#### If add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by XLTechie** [comment]
CC @jcsteh  @josephsl

**Comment by josephsl** [comment]
Hi, actually, Quentin, @feerrenrut, @MichaelDCurran

**Comment by feerrenrut** [comment]
Just to be sure @XLTechie, can you confirm that the anchor in the URL is spelled the same as the `a` name in the HTML file?

If I paste the following URL into Firefox, it opens to the right section:
`https://www.nvaccess.org/files/nvda/documentation/userGuide.html#RunCOMRegistrationFixingTool`

Though we actually write a file with a redirect, because we found that one of the browsers didn't like opening files with anchors from the command line. So the following is the redirect, then encoded to base 64 and written as a dataURL. This works pasted into Firefox.

```
<html><head><meta http-equiv="refresh" content="0;url=https://www.nvaccess.org/files/nvda/documentation/userGuide.html#RunCOMRegistrationFixingTool" /></head></html>
```

```
data:text/html;base64,ZGF0YTp0ZXh0L2h0bWwsPGh0bWw+PGhlYWQ+PG1ldGEgaHR0cC1lcXVpdj0icmVmcmVzaCIgY29udGVudD0iMDt1cmw9aHR0cHM6Ly93d3cubnZhY2Nlc3Mub3JnL2ZpbGVzL252ZGEvZG9jdW1lbnRhdGlvbi91c2VyR3VpZGUuaHRtbCNSdW5DT01SZWdpc3RyYXRpb25GaXhpbmdUb29sIiAvPjwvaGVhZD48L2h0bWw+Cg==
```

```
data:text/html,<html><head><meta http-equiv="refresh" content="0;url=https://www.nvaccess.org/files/nvda/documentation/userGuide.html#RunCOMRegistrationFixingTool" /></head></html>

**Comment by XLTechie** [comment]
I haven't tested this again tonight, but that was the exact spelling and 
capitalization I used.

The refresh URL you gave was an https URI, but 
gui.contextHelp.writeRedirect() generates a file URI (I.E. works off 
line), which it derives from globals.

I am not, however, talking about pasting into Firefox.
What happens if you paste that URL into a run dialog (The file based URL, 
not the https one) with Firefox set as default browser on system?

My results are as described in the issue.


Personally I would like to rewrite context help to provide only the 
appropriate section, in its own ui.browseableMessage, but don't know if a 
PR would be welcome on that subject.
Not related to this issue, although it would certainly solve it.

**Comment by josephsl** [comment]
Hi, as for a browse mode message with just the section content, that’s an extension of #2699, and I think that ought to be the “more useful” user experience. Displaying only the section content also allows me o prepare another add-on of mine to come to Core: Control Usage Assistant. Thanks.

**Comment by XLTechie** [comment]
@josephsl I had been waiting until the end of #11006 to propose it, but I 
couldn't resist mentioning it here, as it would definitely solve this 
issue.

If this issue is real, it's also probably the only way worth pursuing to 
solve it from our side. But I would like someone else to confirm this 
isn't some quirk unique to my setup.

**Comment by XLTechie** [comment]
@feerrenrut were you able to repro this?

**Comment by feerrenrut** [comment]
Firefox was the browser that we couldn't get to work with anchors directly on the file URL. So none of the anchors mentioned work from the run dialog for me. Do they work for you?

The reason I pasted in those data URI tests was to try to narrow down the cause of the issue. If those work, we can be more certain it is related to the local file system. I believe I have a long paths setting enabled in Windows which may give me different behavior than you. 

Just to confirm with Firefox as my default browser, I have tested pressing F1 in the check for updates dialog. It successfully takes me to anchor `#GeneralSettingsCheckForUpdates`.

**Comment by Adriani90** [comment]
@XLTechie can you reproduce this also with NVDA last alpha and firefox 123?

**Comment by XLTechie** [comment]
Yes. Just tested by pressing F1 on "Use UI Automation to access Microsoft Word 
document controls" in Advanced.
The file opened in Firefox, but at the top of the table of contents.

Note that this is now Windows 11 instead of 10, Firefox 123, and a completely 
different computer than when I submitted the issue.

@jcsteh this may be a Firefox bug in regards to file:// URIs with deep 
relative anchors. (I have an FF extension for clearing visited status)

The annoying thing for reproduction, is that once it works, it continues working 
thereafter, even if you clear the "visited" status from within Firefox and 
close the tab.

So my local reproduction targets for this are dwindling.

**Comment by Adriani90** [comment]
@XLTechie, @jcsteh is there a bug filled with Bugzilla already on this one?

**Comment by Adriani90** [comment]
This is still an issue in NVDA 2025.1 Beta. Following is written to the log when I press f1 on a setting while having Firefox as default browser:

```
IO - inputCore.InputManager.executeGesture (01:29:34.397) - winInputHook (3648):
Input: kb(laptop):f1
DEBUG - gui.contextHelp.showHelp (01:29:34.397) - MainThread (21576):
Opening help: helpId = MSWordUIA, userGuidePath: C:\Program Files (x86)\NVDA\documentation\de\userGuide.html
DEBUG - NVDAObjects.IAccessible.chromium.Document._get_treeInterceptorClass (01:29:35.089) - MainThread (21576):
loadChromiumVBufOnBusyState feature flag is True, vBuf WILL be loaded when state of the document is busy.
DEBUG - treeInterceptorHandler.update (01:29:35.090) - MainThread (21576):
Adding new treeInterceptor to runningTable: <NVDAObjects.IAccessible.chromium.ChromeVBuf object at 0x067FF730>
DEBUG - virtualBuffers.VirtualBuffer._loadBuffer (01:29:35.539) - NVDAObjects.IAccessible.chromium.VirtualBuffer.loadBuffer (18832):
Buffer load took 0.447 sec, 363458 chars
DEBUG - watchdog._waitUntilNormalCoreAliveTimeout (01:29:35.593) - watchdog (16724):
Potential freeze, waiting up to 10 seconds.
ERROR - eventHandler.executeEvent (01:29:37.107) - MainThread (21576):
error executing event: gainFocus on <NVDAObjects.IAccessible.chromium.Document object at 0x067EB370> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 348, in executeEvent
  File "eventHandler.pyc", line 399, in doPreGainFocus
  File "NVDAObjects\__init__.pyc", line 374, in objectInForeground
  File "NVDAObjects\__init__.pyc", line 267, in findBestAPIClass
  File "NVDAObjects\__init__.pyc", line 266, in findBestAPIClass
  File "NVDAObjects\UIA\__init__.pyc", line 1515, in kwargsFromSuper
_ctypes.COMError: (-2146233083, None, (None, None, None, 0, None))
DEBUG - browseMode.BrowseModeDocumentTreeInterceptor._getInitialCaretPos (01:29:37.117) - MainThread (21576):
No saved caret position for file:///C:/Program%20Files%20(x86)/NVDA/documentation/de/userGuide.html#MSWordUIA
DEBUG - watchdog._waitUntilNormalCoreAliveTimeout (01:29:37.595) - watchdog (16724):
Recovered from potential freeze after 2.502103800070472 seconds.
DEBUG - watchdog._waitUntilNormalCoreAliveTimeout (01:29:37.622) - watchdog (16724):
Potential freeze, waiting up to 10 seconds.
IO - inputCore.logTimeSinceInput (01:29:37.622) - MainThread (21576):
3.225 sec since input
IO - speech.speech.speak (01:29:37.622) - MainThread (21576):
Speaking ['NVDA alpha-36213,ed59e0e4 Benutzerhandbuch', 'Dokument', CancellableSpeech (still valid)]
IO - inputCore.logTimeSinceInput (01:29:37.625) - MainThread (21576):
3.228 sec since input
IO - speech.speech.speak (01:29:37.625) - MainThread (21576):
Speaking ['Überschrift', 'Ebene 1', 'NVDA alpha-36213,ed59e0e4 Benutzerhandbuch']
DEBUG - watchdog._waitUntilNormalCoreAliveTimeout (01:29:38.123) - watchdog (16724):
Recovered from potential freeze after 1.0008742000209168 seconds.
IO - inputCore.InputManager.executeGesture (01:29:41.742) - winInputHook (3648):
Input: kb(laptop):upArrow
```

**Comment by amirsol81** [comment]
This is also duplicable with Google Chrome V151 and the most recent alpha and official releases of NVDA (Alpha 57522 at the time of this writing).
I get the following error in log:
File "NVDAObjects\UIA\__init__.pyc", line 1185, in _getUIACacheablePropertyValue
_ctypes.COMError: (-2147220991, 'An event was unable to invoke any of the subscribers', (None, None, None, 0, None))
@Qchristensen
